rmw_cyclonedds_cpp/src/serdata.cpp

```diff
@@ -143,9 +143,17 @@ static void serialize_into_serdata_rmw_on_demand(serdata_rmw * d)
   {
     std::lock_guard<std::mutex> lock(type->serialize_lock);
     if (d->iox_chunk && d->data() == nullptr) {
-      serialize_into_serdata_rmw(
-        const_cast<serdata_rmw *>(d),
-        SHIFT_PAST_ICEORYX_HEADER(d->iox_chunk));
+        auto iox_header = iceoryx_header_from_chunk(d->iox_chunk);
+        // if the iox chunk has the data in serialized form
+        if (iox_header->shm_data_state == IOX_CHUNK_CONTAINS_SERIALIZED_DATA) {
+          d->resize(iox_header->data_size);
+          memcpy(d->data(), d->iox_chunk, iox_header->data_size);
+          // TODO(Sumanth), free chunk and set to null?
```

Comment:
I think the chunk should be freed to avoid copying it again to `d->data()` (can this happen?). Otherwise it will be freed when `d` is freed.

The function signature (and name) does not fit the actual semantics, essentially source and destination are both `d` in a way, it is an internal transformation of `d` where we take the data from shared memory and move it into another buffer.

Reply:
It will not happen because `d->data()` will not be null in the next iteration, and I guess the chunk can still be used based on the data path even if the chunk is serialized, so probably we shouldn't free the chunk here yet. \
@eboasson is this correct?